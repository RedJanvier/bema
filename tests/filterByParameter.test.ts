import './__setup'

describe('parameter only filters on value', () => {
  test('only-parameter filters skips a case with no matching parameters', async () => {
    bema
      .settings({ onlyParameters: [':2'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })

  test('multiple only-parameter filters skips a case with one matching among multiple parameters', async () => {
    bema
      .settings({ onlyParameters: [':2', ':1'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('multiple only-parameter filters keeps a case with multiple matching among multiple parameters', async () => {
    bema
      .settings({ onlyParameters: [':1', ':2'] })
      .parameter('p1')
      .parameter('p2')
      .group('a')
      .case({ p1: '1', p2: '2' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(1)
  })
})

describe('parameter skip filters on value', () => {
  test('skip-parameter filter does not skip a case with no matching parameter', async () => {
    bema
      .settings({ skipParameters: [':2'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(1)
  })
  test('skip-parameter filter skips a case with a matching parameter', async () => {
    bema
      .settings({ skipParameters: [':1'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('multiple skip-parameter filters skips a case with one matching among multiple parameters', async () => {
    bema
      .settings({ skipParameters: [':1', { value: /3/ }] })
      .parameter('p1')
      .parameter('p2')
      .group('a')
      .case({ p1: '1', p2: '2' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('multiple skip-parameter filters skips a case with multiple matching among multiple parameters', async () => {
    bema
      .settings({ skipParameters: [':1', ':2'] })
      .parameter('p1')
      .parameter('p2')
      .group('a')
      .case({ p1: '1', p2: '2' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
})

describe('combined skip & only parameter filters', () => {
  test('only-parameter + skip-parameter filters skips a case with no matching only and matching skip parameters', async () => {
    bema
      .settings({ onlyParameters: [{ value: /3/ }], skipParameters: [':1'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('only-parameter + skip-parameter filters skips a case with no matching only and no matching skip parameters', async () => {
    bema
      .settings({ onlyParameters: [{ value: /3/ }], skipParameters: [{ value: /3/ }] })
      .parameter('p1')
      .parameter('p2')
      .group('a')
      .case({ p1: '1', p2: '2' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('only-parameter + skip-parameter filters skips a case with matching only and matching skip parameters', async () => {
    bema
      .settings({ onlyParameters: [':1'], skipParameters: [':1'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(0)
  })
  test('only-parameter + skip-parameter filters keeps a case with matching only and no matching skip parameters', async () => {
    bema
      .settings({ onlyParameters: [':1'], skipParameters: [':2'] })
      .parameter('p1')
      .group('a')
      .case({ p1: '1' })
      .run(f1)
    await run(bema)
    expect(f1.mock.calls.length).toEqual(1)
  })
})

test('throws an error if all tests are filtered out', () => {
  const b = bema.settings({ skipParameters: [':1'], strictMode: true }).parameter('p1')
  b.group('a').case({ p1: '1' })
  expect(run(bema)).rejects.toMatchSnapshot()
})
