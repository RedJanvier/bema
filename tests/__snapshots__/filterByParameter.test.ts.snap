// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws an error if all tests are filtered out 1`] = `
[LintError: [31mAll cases filtered out by parameter filters! This is probably not what you intended.[39m

[2mNote: Multiple filters use AND semantic.[22m

[1m"Only" Filters:[22m
  
    NA

[1m"Skip" Filters:[22m

    â†’ value: /^1$/i

[1mThe available parameter types are:[22m

    â†’ p1

[1mThe parameters of each case are:[22m

    â†’ p1:1]
`;
